pallets/messages/src/lib.rs

```diff
@@ -146,6 +149,21 @@ pub mod pallet {
 			let msa_id = T::AccountProvider::get_msa_id(&who);
 			ensure!(msa_id.is_some(), Error::<T>::InvalidMessageSourceAccount);
 
+			let message_sender_msa = msa_id.unwrap();
+			match on_behalf_of {
+				Some(producer) => {
+					let current_provider = Provider(message_sender_msa);
+					let current_delegator = Delegator(producer);
+					let provider_info = T::AccountProvider::get_provider_info_of(
```

Comment:
maybe move this into a function inside of msa pallet that way you dont have to use Provider and Delegator here. I think this something we will be using often.

```
T::AccountProvider::ensure_valid_provider(provider_msa_id.into(), producer_msa_id.into())
```

Reply:
Since at runtime we are passing MSA singleton as AccountProvider it needs to be added there, however i believe this is still not clean because it depends on #20, So will require some refactoring at that point,  added a task in #140